### PR TITLE
HOST_PROMPT_SUPPORT

### PR DESCRIPTION
**Pull request by InsanityAutomation**
Beware the above code, I have only proved it correct, not tried it. - Knuth
In other words, it compiles. Should be complete enough for some testing at this point. As soon as I get back to my machines Ill test a good bit myself, but any feedback prior is appreciated!

-  M876 implemented in the emergency parser

- M412 extended with an explicit reset option.

- M412 extended with H option for Host or Local handling

-  If the emergency parser is off, M876 is processed in the normal queue

- Capability report for host prompts output

- Prompts on every point of pause / resume with handlers

### CODE REVIEW

File: Marlin/src/feature/emergency_parser.h [hunk 1]
```diff
@@ -114,6 +130,35 @@ class EmergencyParser {
...
+        break;
+
+      case EP_M87:
+        state = (c == '6') ? EP_M876 : EP_IGNORE;
+        break;
+
+      case EP_M876:
+        state = (c == ' ') ? EP_M876S : EP_IGNORE;
+        break;
+
```
Comment (AnHardt) [inline Marlin/src/feature/emergency_parser.h:152]: In g-code white space is optional

**Review by AnHardt — CHANGES_REQUESTED** [review]

File: Marlin/src/feature/emergency_parser.h (hunk 1 above)
Comment (InsanityAutomation) [reply Marlin/src/feature/emergency_parser.h:152]: Good point, ill add a break to the next statement if it sees this and skip it

**Comment by InsanityAutomation** [comment]
@marcio-ao Please note the addition of M412 R here to reset and removal of the automatic reset in runout.h - If the runout distance was set to 0, it would have ended up spamming the trigger so this forces it to explicitly reset from the host side.

Also, instead of not using the defined action when host printing, do you have an objection to defining a second runout action for host vs local printing? I can add an argument to M25 to prevent it from parking by calling M125 then use that for the host printing script, since it would call the defined prompts and action commands for host control.

**Comment by marcio-ao** [comment]
> Please note the addition of M412 R here to reset and removal of the automatic reset in runout.h

It is safe to call "M412 R" when there is not a run-out condition? Right now Cura doesn't necessarily keep track of what caused the pause, so it would need to issue a "M412 R" even if the pause wasn't the result of a filament run-out.

> Also, instead of not using the defined action when host printing, do you have an objection to defining a second runout action for host vs local printing?

I think that would be okay.

**Comment by InsanityAutomation** [comment]
@marcio-ao Sure, it wont do much of anything there, the runout.reset() function will set the ran_out value to what it already is and happily move on with life.

**Comment by InsanityAutomation** [comment]
@marcio-ao My other thought on the above trigger action is to add a handler flag to M412 to toggle between local and host. This way, if a host supports handling and it wants to take control, the connection dialog can include a mode setting command to assume control. If this isnt received, its handled by the runout script. Im a bit concerned users like myself who switch slicers depending on the task (if I didnt use octoprint regardless...) could be burned by it being constant when theyre on a host that doesnt support it.

Edit : sorry to be sortve taking up your whole day going back and forth! We just have alot going on in the same area here and I want to avoid breaking things either legacy or new...

**Comment by marcio-ao** [comment]
> My other thought on the above trigger action is to add a handler flag to M412 to toggle between local and host.

I guess you could do that, but beware that RAM is in short supply and any flag like that ends up being a global variable. Having it a compile time option is preferable in that regards, but as you say, it would make it harder to move from hosts that supported it vs hosts that did not.

**Comment by InsanityAutomation** [comment]
@marcio-ao agreed, every BOOL adds up. In this case, its probably the best way forward to maintain compatibility. Any objection to M412 H<bool> 0 (default) local - use runout script; 1 host - send action and return (still reliant on M412 R as debounce)?

**Comment by InsanityAutomation** [comment]
@marcio-ao and do you think I should advertise this capability for host runout control in M115?

**Comment by marcio-ao** [comment]
> do you think I should advertise this capability for host runout control in M115?

@InsanityAutomation: I don't think it necessarily should be me making those decisions. To me, it doesn't matter.

> Any objection to M412 H 0 (default) local - use runout script; 1 host - send action and return (still reliant on M412 R as debounce)?

Again, this does not really matter to me or to our printers.

**Comment by InsanityAutomation** [comment]
@marcio-ao I figured it did as your PR is what made it so the host action was the only one taken when defined, so I was trying to ensure I didn't cause you too much headache changing that up. This is going to essentially pull back out most of what the ACTION_ON_FILAMENT_RUNOUT PR did.

**Comment by marcio-ao** [comment]
> This is going to essentially pull back out most of what the ACTION_ON_FILAMENT_RUNOUT PR did.

@InsanityAutomation: Don't worry too much about that :) I can change Marlin on our end and possibly even change Cura to comply with the new protocols. The reason for that PR was that at the time there was no way for the host to detect filament runout at all, so I had to improvise. I looked though the wiki you put together and overall I like the direction things are going, even if it means we have to adapt a bit :)

**Comment by InsanityAutomation** [comment]
@marcio-ao Sounds good! Between the toggle added above and the reason code being added to the action command output, the end result should be very close to what you were doing!

**Comment by thinkyhead** [comment]
```
git fetch origin
git checkout RemoteStreamerUpdates_Round2
git reset --hard origin/RemoteStreamerUpdates_Round2
```

**Comment by InsanityAutomation** [comment]
@foosel @repetier  @MagoKimbra Care to at least take a glance and kick me if I missed anything here?

Thanks

File: Marlin/src/Marlin.cpp [hunk 2]
```diff
@@ -396,48 +422,51 @@ void disable_all_steppers() {
...
-  void host_response_handler(char c) { host_response_handler((int)strtol(&c, NULL, 1)); }
-  void host_response_handler(int response) {
-    switch(host_prompt_reason)
-    {
+  void host_response_handler(const uint8_t response) {
+    switch (host_prompt_reason) {
       case PROMPT_FILAMENT_RUNOUT_TRIPPED:
-        if(response==0) {
+        if (response == 0)
           advanced_pause_menu_response = ADVANCED_PAUSE_RESPONSE_EXTRUDE_MORE;
```
Comment (MagoKimbra) [inline Marlin/src/Marlin.cpp]: HOST_PROMPT_SUPPORT can be active, but it is not said to have ADVANCED_PAUSE_FEATURE enabled.
You could have FILAMENT_RUNOUT_SENSOR active to have everything managed by the host and not have ADVANCED_PAUSE_FEATURE, this means a series of checks that are missing, but honestly it seems to me all a bit too complex.

Comment (InsanityAutomation) [reply Marlin/src/Marlin.cpp]: This part will likely change, but at the moment this is set to this value only in the scenario where its being used and all those conditions exist, Id like to break the reliance on the LCD stuff, but lets get it working and tested first!

Comment (MagoKimbra) [reply Marlin/src/Marlin.cpp]: Honestly, I thought of it in a different way. I print from HOST (server / octoprint), the firmware detects the filament end and tells it to the host that it should be him to manage everything. So once the command  // action: is sent, it is the host that takes care of making requests to the user, sending to the firmware the same requests to put on the display so that the user can also respond locally.
So instead the firmware is very complicated, because it may or may not have a function if it has that function there are instructions on the display and send the same instructions from the firmware to the host, is likely to make the firmware too large, in addition to the fact that there are a more and more of checks to check!!!
So either enable advanced pauses to manage everything from the display if I do not have a special host, or enable support for the host prompt, and the only thing the firmware needs to do is report the filament, then it is the host that it does everything.

Comment (InsanityAutomation) [reply Marlin/src/Marlin.cpp]: The point is for the firmware to tell the host how it can, then the host to tell firmware how it wants to handle it. This way, it can still behave nicely if there is no host, but if the host is configured to do it, it can request control and handle it. Aim is to make it seamless. Yes, it's creeping up in size a bit but still not an issue on a 2560 even with tons of options enabled.

**Comment by repetier** [comment]
I'm not so into marlin internals that I see what is really going on. I would say if prompt works with Octoprint it is correct. We have currently no prompt support but will add it maybe in next release. Then I can try to simulate a jam and see if it works. When it works on octoprint I then know I have done something wrong.

**Comment by InsanityAutomation** [comment]
@repetier 
> When it works on octoprint I then know I have done something wrong.

Sounds like a plan there at least! Once foosel has a devel snapshot I can install with M876 and have it processed as an emergency parser command I can go a bit more in depth testing everything. Day job still has me stuck out of town for a bit though.

Hopefully we can get one last review by @thinkyhead then just final testing and/or build on top. Removing graphical LCD dependency is still the next step. Will require either emergency parser and host prompt support or graphical lcd for parking.

**Comment by repetier** [comment]
> > When it works on octoprint I then know I have done something wrong.
Ok, sounds not exactly like what I wanted to say :-)
> Sounds like a plan there at least! Once foosel has a devel snapshot I can install with M876 and have it processed as an emergency parser command I can go a bit more in depth testing everything. Day job still has me stuck out of town for a bit though.

As I understood prompt support is already included. Just M876 is not an emergency command since it was designed for non blocking communication. So guess that is the only thing required to change. I have it already as emergency command listed in 0.91.0 just do not show the prompt:-)

**Comment by InsanityAutomation** [comment]
@repetier Yeah, I creatively snipped for comedy! 

I can do some testing when Im back home tomorrow I hope, just not 100% until I can verify emergency parsing as well.

File: Marlin/src/Marlin.cpp [hunk 3]
```diff
@@ -320,28 +320,181 @@ void disable_all_steppers() {
...
+      else {
+        // Legacy Repetier command to be used until newer version support standard dialog
+        // Will be removed at a later date when pause command also triggers dialog
+        SERIAL_ECHOPGM("//action:out_of_filament T");
+        SERIAL_CHAR(tool);
+        SERIAL_EOL();
+
+        host_action_pause(false);
+      }
+      SERIAL_ECHOPGM(" ACTION_REASON_ON_FILAMENT_RUNOUT");
```
Comment (thinkyhead) [inline Marlin/src/Marlin.cpp]: Why is this printing the literal string `" ACTION_REASON_ON_FILAMENT_RUNOUT"` to the host?

Comment (InsanityAutomation) [reply Marlin/src/Marlin.cpp]: Lack of coffee induced typo.....

**Comment by thinkyhead** [comment]
Rebased, squashed, etc. …
```
git fetch origin
git checkout RemoteStreamerUpdates_Round2
git reset --hard origin/RemoteStreamerUpdates_Round2
```

**Comment by InsanityAutomation** [comment]
I'll take a look through the changes most likely tomorrow morning. I'm on the road between Pennsylvania and Michigan for the rest of today.

**Comment by repetier** [comment]
@InsanityAutomation You can already test with Repetier-Server 0.91.0 if you use the console. You will not get a popup but the prompt messages are in log and M876 is a emergency command already in that version. So testing should be no problem.
@MagoKimbra Same for you since you asked me yesterday.

**Comment by MagoKimbra** [comment]
Thank's @repetier

**Comment by thinkyhead** [comment]
Rebased and squashed some commits:
```
git fetch origin
git checkout RemoteStreamerUpdates_Round2
git reset --hard origin/RemoteStreamerUpdates_Round2
```

**Comment by thinkyhead** [comment]
Rebased, squashed, cleaned up, reorganized…
```
git fetch origin
git checkout RemoteStreamerUpdates_Round2
git reset --hard origin/RemoteStreamerUpdates_Round2
```

**Comment by foosel** [comment]
> I'm not so into marlin internals that I see what is really going on.

I'm a bit more into Marlin internals but still don't feel comfortable commenting on the implementation details here right now ;)



> Once foosel has a devel snapshot I can install with M876

The current `maintenance` branch (what is going to be 1.3.11) extends the prompt support that has been in OctoPrint for nearly a year now by making `M876` the configured default command and adding it to the set of commands known to be supported by the emergency parser **if** the firmware reports the emergency parser capability and the `PROMPT_SUPPORT` capability we discussed in the other thread and that I also documented [on the wiki](https://reprap.org/wiki/G-code#M876:_Dialog_handling).

**Comment by InsanityAutomation** [comment]
@foosel sounds good. This has had a bunch of stuff rebased again, so I'll go after flashing the current version and pull a maintenance snapshot on a machine tonight.

**Comment by MagoKimbra** [comment]
@foosel How do I install the maintenance version on windows?
I have installed everything but I think I have installed the master version.

edit: ok i found the path. I install maintenance.
Ok, the prompt function works. Now the problem is the response .. But I think the problem is mine ...

Edit: Ok now everything goes, only one thing at the end I send the command \\action: resumed, but octoprint stays still until I press the resume button. Where am I wrong?

**Comment by foosel** [comment]
> I send the command \action: resumed, but octoprint stays still until I press the resume button. Where am I wrong?

If you really sent it like that, in the command format. It should be `//action:resumed`, not `\action:resumed`.

**Comment by MagoKimbra** [comment]
Yes sorry, i wrote badly i send command //action:resumed

**Comment by foosel** [comment]
Then I need more information in order to tell you what goes wrong here. Please create a `serial.log`. Might be a bug, might be something wrong in the communication, but impossible to say yet.

**Comment by MagoKimbra** [comment]
[serial.log](https://github.com/MarlinFirmware/Marlin/files/2829024/serial.log)

**Comment by foosel** [comment]
Ah, I get it. OctoPrint was still switching to paused state (note that it wasn't yet paused but pausing because it hadn't yet recorded the pause position), which it couldn't though since the communication was blocked the full time. So when it received the resumed command that didn't go through since the actual paused state was only reached after. I'll have to think of something for situations like that. There's actually a ticket somewhere on my bugtracker about a similar situation when the printer is resumed "too fast" for other reasons, causing things to not align properly. Will hopefully only be a case of a slightly different state evaluation.

**Comment by InsanityAutomation** [comment]
@foosel for `paused` and `resumed` where scripts are not executed, should be able to change state immediately without recording the position as it's not needed.

**Comment by foosel** [comment]
@InsanityAutomation there are also events fired in such cases and whatever users have setup with those might break in such cases. But I agree, this is probably a death we'll have to die here. Needs some more intense testing though and might take a few days.

**Comment by InsanityAutomation** [comment]
@foosel I agree it'll probably need to break those. The reason those are sent is to inform when things can't be executed.

**Comment by InsanityAutomation** [comment]
@foosel It would seem the prompts are not parsed when it is in pausing state. We did however get the pausing prompt when parking, then the continue, followed by the reheat as expected. Im not sure which end (the terminal window or machine not accepting) is causing it to not respond to M876 at the moment, so I may do some more testing from a putty terminal tomorrow. It did seem to get a line number so im guessing from the terminal window isnt flagged as emergency parser and still gets a line count.

Im on Commit 61431c0df3fb571bd5fdaa6a910e4366a2145c51

I still havnt been through all of the changes @thinkyhead made, so ive got to sit down and review that as well... Pesky day job getting in the way lol
```
Recv: ok N37 P1 B2
Recv: //action:paused
Send: N39 G1 X302.028 Y305.236 E37.2944*93
```
Printer signalled that it paused, switching state...
Changing monitoring state from "Printing" to "Pausing"
```
Recv: //action:prompt_end
Recv: //action:prompt_begin Pause
Recv: //action:prompt_show
Recv: echo:busy: processing
Recv: echo:busy: processing
Recv: echo:busy: processing
Recv: echo:busy: processing
```
Did not receive parseable position data from printer within 10.0s, continuing without it
```
Recv: echo:busy: processing
Recv: echo:busy: processing
Recv: echo:busy: processing
Recv: echo:busy: processing
Recv: echo:Insert filament and press button (or M108)
Recv: //action:prompt_end
Recv: //action:prompt_begin Paused
Recv: //action:prompt_button PurgeMore
Recv: //action:prompt_button Continue
Recv: //action:prompt_show
Recv: echo:busy: paused for user
Recv: echo:busy: paused for user
Recv: echo:busy: paused for user
Recv: echo:Press button (or M108) to heat nozzle
Recv: //action:prompt_end
Recv: //action:prompt_begin HeaterTimeout
Recv: //action:prompt_button Reheat
Recv: //action:prompt_show
Recv: echo:busy: paused for user
Recv: echo:busy: paused for user
```

**Comment by InsanityAutomation** [comment]
I did get some responses to go through and get some debug info out so I know what to chase on that side now at least. I still havnt seen a prompt dialog show up though

**Comment by foosel** [comment]
> I still havnt seen a prompt dialog show up though

Is the firmware signaling prompt capabilities in the response to `M115`? OctoPrint 1.3.11+ will only process prompts if the corresponding capability is reported by the firmware (or if the user has configured it to ignore capability reports and forced support in the settings). Same for emergency parser handling of `M876`.

**Comment by InsanityAutomation** [comment]
It did have the cap line, I forgot to paste that in the above. I ran out of time last night so I'll have to poke at it some more tonight after I fix the few issues I found on my end.

**Comment by InsanityAutomation** [comment]
@foosel i pulled devel tonight. I have prompts showing up and some answers making it over. It does still cause an issue resuming that the paused state doesn't trigger immediately so resumed is missed. It also appears that the dialog responses are getting queued where I need to fake acknowledgement to get them to make their way to the machine.

**Comment by InsanityAutomation** [comment]
@foosel ive also noticed that the only way to abort a dialog is the escape key, which is not shown on the dialog anywhere. Possible a UI enhancement at some point.

Back to breaking more things.... then sleep.... lol

**Comment by InsanityAutomation** [comment]
@foosel If I use fake acknowledgement generously, everything is working with the devel snapshot OctoPrint: 1.4.0.dev2000+g3c5aa60d 
Installed: Commit 3c5aa60d2b92b67e9c3fa5b587e3d709b0d5a536

I have one last rogue message that came on when the condition had passed to chase down and things should be good on this end. If you think the buffer issue going to continue to cause problems for awhile I can add a config option to clear the buffer of commands on runout before executing m600 after sending the paused state to leave as much as possible open.

**Comment by thinkyhead** [comment]
Rebased, squashed, etc. …
```
git fetch origin
git checkout RemoteStreamerUpdates_Round2
git reset --hard origin/RemoteStreamerUpdates_Round2
```

**Comment by InsanityAutomation** [comment]
IMO this is set. Id like to see a bit wider testing, but im not sure what if any ill be able to get from others before Octoprint 1.3.11 is released. I couldnt reproduce the rogue error I got the other day, and it was most likely a symptom of me still working with fake acknowledgements.

**Comment by foosel** [comment]
@InsanityAutomation just for the record, `devel` != 1.3.11 ;) `maintenance` is where it's at.

I've adjusted the code for 1.3.11 so that 

  * `Pausing` is treated as `Paused` when it comes to handling of `action:resume`, `action:resumed`, the user clicking resume etc.
  * no position capturing takes place on `action:paused`

**Comment by InsanityAutomation** [comment]
@foosel I found that and switched branches - things worked a little better :) 

The adjustment to pausing should help a lot!

File: Marlin/src/lcd/menu/menu_main.cpp [hunk 4]
```diff
@@ -147,8 +158,8 @@ void menu_main() {
       }
     #endif // !HAS_ENCODER_WHEEL && SDSUPPORT
 
-    #if ENABLED(SDSUPPORT) || defined(ACTION_ON_RESUME)
-      #if ENABLED(SDSUPPORT)
+    #if ENABLED(SDSUPPORT) || ENABLED(HOST_ACTION_COMMANDS)
+      #if DISABLED(HOST_ACTION_COMMANDS)
         if (card.isFileOpen() && card.isPaused())
       #endif
           MENU_ITEM(function, MSG_RESUME_PRINT, lcd_resume);
```
Comment (thinkyhead) [inline Marlin/src/lcd/menu/menu_main.cpp:165]: @InsanityAutomation — This needs a proper condition so it only appears when there is a print actually in progress. Other code looks at the print job timer and/or the `planner.movesplanned` value to tell that there is a print in progress. If the job timer is paused, then we surely know the print is paused. And if there are moves in the planner, we may act as though the print is in progress (not paused).

Comment (InsanityAutomation) [reply Marlin/src/lcd/menu/menu_main.cpp:165]: Its inside the else of if(busy) so that should have it covered for the when active / not paused state. For a print in progress, I can add that if the timer feature is enabled. It is by default so in most cases then it will.

**Comment by MagoKimbra** [comment]
My tests with Octorpint 1.3.11 and MK4duo seem to be fine.
I have other tests, everything seems to work.

File: Marlin/src/Marlin.cpp [hunk 5]
```diff
@@ -315,37 +319,98 @@ void disable_all_steppers() {
...
+        host_action_paused(false);
+      }
+      else {
+        // Legacy Repetier command for use until newer version supports standard dialog
+        // To be removed later when pause command also triggers dialog
+        #ifdef ACTION_ON_FILAMENT_RUNOUT
+          host_action(PSTR(ACTION_ON_FILAMENT_RUNOUT " T"), false);
+          SERIAL_CHAR(tool);
+          SERIAL_EOL();
+        #endif
```
Comment (thinkyhead) [inline Marlin/src/Marlin.cpp:368]: Is `ACTION_ON_FILAMENT_RUNOUT` made fully obsolete by this PR?

File: Marlin/src/feature/host_actions.cpp [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "../inc/MarlinConfig.h"
+
+#if ENABLED(HOST_ACTION_COMMANDS)
+
+#include "host_actions.h"
```
Comment (thinkyhead) [inline Marlin/src/feature/host_actions.cpp:27]: The indentation is correct. Don't indent the whole file within the main condition, as this doesn't make for any useful code-folding.

File: Marlin/src/Marlin.cpp (hunk 5 above)
Comment (InsanityAutomation) [reply Marlin/src/Marlin.cpp:368]: Pretty close to, but not quite. People may want to have it do something different or extra tasks on pause / recover. I considered in the next half when I get the rest of the LCD reliance out setting that to default with ifndef to m600/m25 depending on advanced pause being enabled and allowing users to override, but this changes behavior enough already and I didn't want to keep growing this pr.

**Comment by marcio-ao** [comment]
@InsanityAutomation: I am reviewing the changes introduced in this PR, and I do see one potential problem. Currently, you added a "M412 H" command that the host can use to notify Marlin that it will handle the filament change. This is something I would like to use in Cura as part of our start GCODE. The problem I see with the current implementation is that the "M412 H" command also causes Marlin to ignore the script if that that command is found during an SD card print.

The reason this is a problem is that in the Cura start GCODE, I cannot conditionally issue that command based on whether the printing is happening via USB or being saved to a file on the SD card. Cura will always generate the same GCODE regardless.

So to get the desired behavior, the the definition of "M412 H" would need to be changed to "allow the host to handle the filament change when not printing from the SD card". This would mean that "M412 H" would be ignored if encountered while printing from an SD card.

**Comment by InsanityAutomation** [comment]
@marcio-ao If a host is connected and ready to handle the action, is the source of the print relevant? We can add a check for this condition in M25 before it calls M125 to not enter the park function if host is in control.

Potential ways to avoid it if really desired though, parse results of M27 to determine if SD printing

Inject M412 H in the start / end of the print from usb so that mode is only sent when printing from cura, not on connect.

File: Marlin/src/Marlin.cpp (hunk 5 above)
Comment (InsanityAutomation) [reply Marlin/src/Marlin.cpp:368]: I noticed I misread this yesterday. The above would only be used for driving the current dialog in repetier. When theyre dialog matches octoprint with everything else it can be deprecated out. Its a compatibility element atm.

**Comment by marcio-ao** [comment]
> If a host is connected and ready to handle the action, is the source of the print relevant?

@InsanityAutomation: As it stands, Cura does not write different GCODE when it is connected and printing to USB vs. when it is not connected and saving GCODE to an SD card. So the way I had worked around this problem before was having Marlin execute the runout script only when not printing from the SD card.

> Inject M412 H in the start / end of the print from usb so that mode is only sent when printing from cura, not on connect.

We could modify Cura to do this, but there are corner cases that could be missed. What if the user prints from Cura, but then accidentally reboots his computer? Now the user may immediately decide to redo the print from SD instead without Cura (to avoid the same thing from happening again). But now, since the previous Cura print never finished, the printer will remember the "M412 H" and will fail to detect a runout if it happens during the SD card print. The user just ended up with two failed prints in a row. Not good.

How about the following change:

```
const bool marlin_handles_runout = IS_SD_PRINTING() || !runout.host_handling;

#if ENABLED(HOST_ACTION_COMMANDS)
      if (marlin_handles_runout
        && ( strstr(FILAMENT_RUNOUT_SCRIPT, "M600")
          || strstr(FILAMENT_RUNOUT_SCRIPT, "M125")
          #if ENABLED(ADVANCED_PAUSE_FEATURE)
            || strstr(FILAMENT_RUNOUT_SCRIPT, "M25")
          #endif
        )
      ) {
        host_action_paused(false);
      }
      else {
        // Legacy Repetier command for use until newer version supports standard dialog
        // To be removed later when pause command also triggers dialog
        #ifdef ACTION_ON_FILAMENT_RUNOUT
          host_action(PSTR(ACTION_ON_FILAMENT_RUNOUT " T"), false);
          SERIAL_CHAR(tool);
          SERIAL_EOL();
        #endif

        host_action_pause(false);
      }
      SERIAL_ECHOPGM(" " ACTION_REASON_ON_FILAMENT_RUNOUT " ");
      SERIAL_CHAR(tool);
      SERIAL_EOL();
    #endif // HOST_ACTION_COMMANDS

    if (marlin_handles_runout)
      enqueue_and_echo_commands_P(PSTR(FILAMENT_RUNOUT_SCRIPT));
```

This would mean that the host could request to handle runout using "M425 H", but that election would only take place if not printing from the SD card.

**Comment by InsanityAutomation** [comment]
@marcio-ao I havnt printed from USB on my pc in so long, I didnt even think about the win10 update randomness that tends to shut people down... What about a basic host timeout that resets the state to internal? Can store the printduration value if the time is running and the rx buffer is not empty. If the buffer is empty and the difference gets over X we reset the host state. Does cura still send any keep alive statements when receiving busy during heating and such? With this, we could possibly save cpu cycles when no host is connected by just binning the output unless some activity has been seen by the host.

May be a whole new ball of wax to do that though. Going back to the statement from earlier, this would be a case of the machine doing its own thing instead of what the host explicitly instructs it to do. The host asked for that control and everything that goes with it. 

We could also reset the state on M84, so its cleared when a print is stopped, but that doesnt help if youre not injecting commands at start / stop.

TL;DR I dont really like just ignoring the explicit request by the host to handle it, but I dont necessarily see another quick easy solution.

**Comment by marcio-ao** [comment]
> What about a basic host timeout that resets the state to internal? Can store the printduration value if the time is running and the rx buffer is not empty. If the buffer is empty and the difference gets over X we reset the host state.

@InsanityAutomation: I feel all that would over complicate things. It's clear that the way forward is simply for Cura LE to implement HOST_PROMPT_SUPPORT and I'll push for that being done soon.

For the meanwhile, I just need to modify our FW to have the old behavior so it is compatible with the existing Cura LE. I am pushing a new PR that fixes a compile error in the code and adds a const variable to make it slightly easier for me to do the modification. No biggie, all I did was add a const variable, see #13163.

**Comment by marcio-ao** [comment]
I plan to modify our FW like this:

```
    #if defined(LULZBOT_RUNOUT_HANDLING_WORKAROUND)
      const bool run_runout_script = IS_SD_PRINTING();
      if(!IS_SD_PRINTING())
        runout.reset();
    #else
      const bool run_runout_script = !runout.host_handling;
    #endif
```

This restores the previous behavior, at least until we can get a better Cura LE out there. You can now see why I wanted that const variable in there.
